### PR TITLE
fix: send Content-Length header if there is payload

### PR DESCRIPTION
**Pull request by camelmasa**
Fixed: https://github.com/coreruleset/go-ftw/issues/134

CRS 942101-* tests received "411 Length Required" data. go-ftw sends Content-Length header if request data is not 0 byte for now.
https://github.com/coreruleset/go-ftw/blob/main/ftwhttp/header.go#L78-L80

I think the Content-Length header needed if request header's method is "POST".

### CODE REVIEW

File: ftwhttp/request.go [hunk 1]
```diff
@@ -103,8 +104,14 @@ func (r *Request) AddHeader(name string, value string) {
 // AddStandardHeaders adds standard headers to the request, if they don't exist
 //
 // This will add Content-Length and the proper Content-Type
-func (r *Request) AddStandardHeaders(size int) {
-	r.headers.AddStandard(size)
+func (r *Request) AddStandardHeaders() {
+	// For better performance, we always close the connection (unless otherwise)
```
Comment (theseion) [inline ftwhttp/request.go:108]:
```suggestion
```

File: ftwhttp/request.go [hunk 2]
```diff
@@ -103,8 +104,14 @@ func (r *Request) AddHeader(name string, value string) {
...
+	r.headers.Add("Connection", "close")
```
Comment (theseion) [inline ftwhttp/request.go:109]: Set only if the `Connection` header isn't already present.

File: ftwhttp/request.go [hunk 3]
```diff
@@ -103,8 +104,14 @@ func (r *Request) AddHeader(name string, value string) {
...
+
+	// If there is data or POST method header, we add the length also
+	if len(r.data) > 0 || r.requestLine.Method == "POST" {
```
Comment (theseion) [inline ftwhttp/request.go:112]: Define `methodsWithBodyRegex` as a constant with `regexp.MustCompile("^POST|PUT|PATCH|DELETE$")`.

```suggestion
	if len(r.data) > 0  || methodsWithBodyRegex.matchString(r.requestLine) {
```

File: ftwhttp/request.go [hunk 4]
```diff
@@ -103,8 +104,14 @@ func (r *Request) AddHeader(name string, value string) {
 // AddStandardHeaders adds standard headers to the request, if they don't exist
 //
 // This will add Content-Length and the proper Content-Type
-func (r *Request) AddStandardHeaders(size int) {
-	r.headers.AddStandard(size)
+func (r *Request) AddStandardHeaders() {
+	// For better performance, we always close the connection (unless otherwise)
+	r.headers.Add("Connection", "close")
+
+	// If there is data or POST method header, we add the length also
```
Comment (theseion) [inline ftwhttp/request.go:111]:
```suggestion
```

File: ftwhttp/request.go [hunk 5]
```diff
@@ -103,8 +104,14 @@ func (r *Request) AddHeader(name string, value string) {
 // AddStandardHeaders adds standard headers to the request, if they don't exist
 //
 // This will add Content-Length and the proper Content-Type
```
Comment (theseion) [inline ftwhttp/request.go:106]:
```suggestion
// AddStandardHeaders does the following:
//   - adds `Connection` header with `close` value (if not set) to improve performance
//   - adds `Content-Length` header if payload size > 0 or the request method
//      permits a body (the spec says that the client SHOULD send `Content-Length`
//      in that case)
```

**Review by theseion — CHANGES_REQUESTED** [review]
Thanks!

**Comment by camelmasa** [comment]
@theseion Got it. Thanks for reviewing. I'll update them then.

File: ftwhttp/request.go (hunk 2 above)
Comment (camelmasa) [reply ftwhttp/request.go]: Updated by https://github.com/coreruleset/go-ftw/pull/135/commits/84028c84b876ff6c60490572a97c5a946d12868c

File: ftwhttp/request.go (hunk 3 above)
Comment (camelmasa) [reply ftwhttp/request.go]: Updated by https://github.com/coreruleset/go-ftw/pull/135/commits/a6e2e79d88d9b9d7d6904bd5c131a9b7aab54e6a

**Comment by camelmasa** [comment]
@theseion Okay, I updated them. How is this?

File: ftwhttp/request.go [hunk 6]
```diff
@@ -102,9 +104,23 @@ func (r *Request) AddHeader(name string, value string) {
...
+		r.headers.Add("Connection", "close")
+	}
+
+	if len(r.data) > 0 || methodsWithBodyRegex().MatchString(r.requestLine.Method) {
+		r.headers.Add("Content-Length", strconv.Itoa(len(r.data)))
+	}
+}
+
+func methodsWithBodyRegex() *regexp.Regexp {
+	return regexp.MustCompile(`^POST|PUT|PATCH|DELETE$`)
```
Comment (theseion) [inline ftwhttp/request.go:123]: I'd rather have this defined as a constant at file scope. It's not a huge deal but this regex will be applied many times during a full test run. Thus I'd like to cache compilation.

```go
var methodsWithBodyRegex = regexp.MustCompile(`^POST|PUT|PATCH|DELETE$`)
```

**Review by theseion — CHANGES_REQUESTED** [review]

**Comment by theseion** [comment]
Thanks! Almost perfect.

File: ftwhttp/request.go (hunk 6 above)
Comment (camelmasa) [reply ftwhttp/request.go:123]: I see! Good catch. 👍

Comment (camelmasa) [reply ftwhttp/request.go]: @theseion Updated https://github.com/coreruleset/go-ftw/pull/135/commits/4bfdaca220be250a5e00860177778608ace60aad

**Comment by jcchavezs** [comment]
> That's great, thanks. Would you mind adding 6 tests for that change? One for each of the methods with body, one for GET, and one for GET + non-empty payload (this should set the Content-Length header regardless of the method)?

Does it make a difference actually? I would just go for GET+e.g.PUT with/without body and move the regex verb check into a meaningful private method which could be tests if we really want to get 100% coverage.

Side note: What about `OPTIONS` verb?

**Comment by theseion** [comment]
> Does it make a difference actually (to add 6 tests)? I would just go for GET+e.g.PUT with/without body and move the regex verb check into a meaningful private method which could be tests if we really want to get 100% coverage. I am coming from the fact that we should test in isolation as much as possible and testing all 4 verbs does not seem to add more value.

I'm not a Go expert, so I'm not very familiar with idioms and best practices. My reasoning for testing each verb is that I want to know when one of them disappears. There may also come a point where we have to extend the list (e.g. for WebDAV verbs) and then I'd want an additional test for each verb.
I'm not looking for 100% coverage, that doesn't add much value, IMO :).

> Side note: What about OPTIONS verb?

`OPTIONS` has no body, so the `Content-Length` header is unnecessary.

**Comment by camelmasa** [comment]
@theseion I added tests https://github.com/coreruleset/go-ftw/pull/135/commits/e7b9771cb4ff9cafcdf8d9692fa5c60cf9cae580.

Also, I'm not Go expert too 😓

**Review by theseion — CHANGES_REQUESTED** [review]
Looks good.
Can you add one last test? I want to make sure that it is possible to send data in a `GET` request. So the request method should be `GET`, the payload should not be empty and that should force the `Content-Length` header to be set.

**Comment by camelmasa** [comment]
@theseion Okay, added tests https://github.com/coreruleset/go-ftw/pull/135/commits/f496432ca7878c3250b9525ddf2ed9eb96a06b8c

**Review by theseion — APPROVED** [review]

**Comment by theseion** [comment]
Thanks for all the work @camelmasa!

**Comment by camelmasa** [comment]
Thanks for reviewing!
